File: frontend/src/app/services/admin-service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../enviroments/environment';
import { ConfesionAdminDTO } from '../models/confesiones.dtos';

@Injectable({
  providedIn: 'root'
})
export class AdminService {
  private apiUrl = `${environment.apiUrl}/admin`;

  constructor(private http: HttpClient) { }

  private getHeaders(usuarioId: number): HttpHeaders {
    return new HttpHeaders({
      'Usuario-Id': usuarioId.toString()
    });
  }

  // Obtener todas las confesiones con autor (solo admin)
  getConfesionesConAutor(usuarioId: number): Observable<ConfesionAdminDTO[]> {
    return this.http.get<ConfesionAdminDTO[]>(`${this.apiUrl}/confesiones`, {
      headers: this.getHeaders(usuarioId)
    });
  }

  // Eliminar confesión (soft delete)
  eliminarConfesion(usuarioId: number, confesionId: number): Observable<{ mensaje: string }> {
    return this.http.delete<{ mensaje: string }>(
      `${this.apiUrl}/confesiones/${confesionId}`,
      { headers: this.getHeaders(usuarioId) }
    );
  }

  // Revelar autor de una confesión
  revelarAutor(usuarioId: number, confesionId: number): Observable<{ mensaje: string }> {
    return this.http.put<{ mensaje: string }>(
      `${this.apiUrl}/confesiones/${confesionId}/revelar`,
      {},
      { headers: this.getHeaders(usuarioId) }
    );
  }

  // Banear usuario
  banearUsuario(usuarioId: number, usuarioABanearId: number): Observable<{ mensaje: string }> {
    return this.http.put<{ mensaje: string }>(
      `${this.apiUrl}/usuarios/${usuarioABanearId}/banear`,
      {},
      { headers: this.getHeaders(usuarioId) }
    );
  }

  // Desbanear usuario
  desbanearUsuario(usuarioId: number, usuarioADesbanearId: number): Observable<{ mensaje: string }> {
    return this.http.put<{ mensaje: string }>(
      `${this.apiUrl}/usuarios/${usuarioADesbanearId}/desbanear`,
      {},
      { headers: this.getHeaders(usuarioId) }
    );
  }

}

File: src/main/java/com/example/Confesionario/service/AdminService.java
package com.example.Confesionario.service;


import com.example.Confesionario.repository.ComentarioRepository;
import com.example.Confesionario.repository.ConfesionRepository;

import com.example.Confesionario.dto.confesion.ConfesionAdminDTO;
import com.example.Confesionario.entities.Confesion;
import com.example.Confesionario.entities.Usuario;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import com.example.Confesionario.repository.UsuarioRepository;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class AdminService {

    @Autowired
    private ConfesionRepository confesionRepository;

    @Autowired
    private UsuarioRepository usuarioRepository;

    @Autowired
    private ComentarioRepository comentarioRepository;

    // Verificar si un usuario es admin
    public boolean esAdmin(Long usuarioId) {
        return usuarioRepository.findById(usuarioId)
                .map(Usuario::getEsAdmin)
                .orElse(false);
    }

    // Obtener todas las confesiones CON información del autor real
    public List<ConfesionAdminDTO> getConfesionesConAutor() {
        List<Confesion> confesiones = confesionRepository.findAll();

        return confesiones.stream()
                .map(c -> {
                    int comentarios = (int) comentarioRepository.countByConfesionId(c.getId());
                    return new ConfesionAdminDTO(c);
                })
                .collect(Collectors.toList());
    }

    // Eliminar confesión (soft delete)
    @Transactional
    public void eliminarConfesion(Long confesionId) {
        Confesion confesion = confesionRepository.findById(confesionId)
                .orElseThrow(() -> new RuntimeException("Confesión no encontrada"));

        confesion.setEliminada(true);
        confesionRepository.save(confesion);
    }

    // Revelar autor de una confesión
    @Transactional
    public void revelarAutor(Long confesionId) {
        Confesion confesion = confesionRepository.findById(confesionId)
                .orElseThrow(() -> new RuntimeException("Confesión no encontrada"));

        confesion.setRevelada(true);
        confesionRepository.save(confesion);
    }

    // Banear usuario
    @Transactional
    public void banearUsuario(Long usuarioId) {
        Usuario usuario = usuarioRepository.findById(usuarioId)
                .orElseThrow(() -> new RuntimeException("Usuario no encontrado"));

        usuario.setBaneado(true);
        usuarioRepository.save(usuario);
    }

    // Desbanear usuario
    @Transactional
    public void desbanearUsuario(Long usuarioId) {
        Usuario usuario = usuarioRepository.findById(usuarioId)
                .orElseThrow(() -> new RuntimeException("Usuario no encontrado"));

        usuario.setBaneado(false);
        usuarioRepository.save(usuario);
    }
}